File: src/main/kotlin/com/xs0/gqlktx/schema/intro/GqlIntroType.java
package com.xs0.gqlktx.schema.intro;

import com.xs0.gqlktx.ann.GQLArg;
import com.xs0.gqlktx.ann.GraphQLObject;
import com.xs0.gqlktx.schema.builder.TypeKind;
import com.xs0.gqlktx.types.gql.*;

import java.util.List;

@GraphQLObject("__Type")
public class GqlIntroType {
    private final GType type;

    public GqlIntroType(GType type) {
        this.type = type;
    }

    public TypeKind getKind() {
        return type.getKind();
    }

    public String getName() {
        if (!(type instanceof GBaseType))
            return null;

        return ((GBaseType)type).getName();
    }

    public String getDescription() {
        return null; // TODO
    }

    public List<GqlIntroField> getFields(@GQLArg(defaultsTo = "false") boolean includeDeprecated) {
        if (!(type instanceof GFieldedType))
            return null;

        return ((GFieldedType)type).getFieldsForIntrospection(includeDeprecated);
    }

    public List<GqlIntroType> getInterfaces() {
        if (type.getKind() != TypeKind.OBJECT)
            return null;

        return ((GObjectType) type).getInterfacesForIntrospection();
    }

    public List<GqlIntroType> getPossibleTypes() {
        if (type.getKind() != TypeKind.UNION)
            return null;

        return ((GUnionType) type).getMembersForIntrospection();
    }

    public List<GqlIntroEnumValue> getEnumValues(@GQLArg(defaultsTo = "false") boolean includeDeprecated) {
        if (type.getKind() != TypeKind.ENUM)
            return null;

        return ((GEnumType)type).getValuesForIntrospection(includeDeprecated);
    }

    public List<GqlIntroInputValue> getInputFields() {
        if (type.getKind() != TypeKind.INPUT_OBJECT)
            return null;

        return ((GInputObjType)type).getInputFieldsForIntrospection();
    }

    public GqlIntroType getOfType() {
        if (!(type instanceof GWrappingType))
            return null;

        return ((GWrappingType)type).getWrappedType().introspector();
    }
}

File: src/main/kotlin/com/xs0/gqlktx/schema/intro/GqlIntroType.kt
package com.xs0.gqlktx.schema.intro

import com.xs0.gqlktx.GqlParam
import com.xs0.gqlktx.GqlObject
import com.xs0.gqlktx.schema.builder.TypeKind
import com.xs0.gqlktx.types.gql.*

@GqlObject("__Type")
class GqlIntroType(private val type: GType) {

    val kind: TypeKind
        get() = type.kind

    val name: String?
        get() = (type as? GBaseType)?.name

    val description: String?
        get() = (type as? GBaseType)?.description

    fun getFields(@GqlParam(defaultsTo = "false") includeDeprecated: Boolean): List<GqlIntroField>? {
        return (type as? GFieldedType)?.getFieldsForIntrospection(includeDeprecated)
    }

    val interfaces: List<GqlIntroType>?
        get() = (type as? GObjectType)?.interfacesForIntrospection

    val possibleTypes: List<GqlIntroType>?
        get() = when (type) {
            is GInterfaceType -> type.implsForIntrospection
            is GUnionType -> type.membersForIntrospection
            else -> null
        }

    fun getEnumValues(@GqlParam(defaultsTo = "false") includeDeprecated: Boolean): List<GqlIntroEnumValue>? {
        return (type as? GEnumType)?.getValuesForIntrospection(includeDeprecated)
    }

    val inputFields: List<GqlIntroInputValue>?
        get() = if (type.kind != TypeKind.INPUT_OBJECT) null else (type as GInputObjType).inputFieldsForIntrospection

    val ofType: GqlIntroType?
        get() = (type as? GWrappingType)?.wrappedType?.introspector
}
